frontend/src/metabase/collections/components/CollectionHeader/CollectionHeader.tsx

```diff
@@ -35,16 +36,25 @@ const CollectionHeader = ({
   onDeleteBookmark,
   onUpload,
   canUpload,
+  uploadsEnabled,
 }: CollectionHeaderProps): JSX.Element => {
+  const showUploadButton =
+    collection.can_write && (canUpload || !uploadsEnabled);
```

Comment:
the tricky logic that this filters out is where a user that has write permission on the collection, uploads are enabled for the instance, but the user does not have data access to the db.